New NXP ENET driver

decsny opened the pull request:
ETH MCUX driver is unmaintainable and cannot support a PHY other than the one on some of the NXP EVKs. Redo, planning to deprecate old driver but keep for a release or two.

New driver selects `EXPERIMENTAL` until the community has had a chance to try this new driver.

This PR also makes the phy-connection-type property generic to all ethernet controllers

Commit:
dts: bindings: Add NXP ENET bindings

Add bindings for compatibles related to NXP ENET IP.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
include: NXP ENET driver include header

Create an include header to be used by NXP ENET drivers.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: mdio: Add NXP ENET MDIO driver

Add driver for NXP ENET MDIO functionalities

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: clock_control: mcux_ccm: Add ENET clock

Add ENET clock value to the CCM clock decoder
for both RT10XX and RT11XX CCM versions.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: ethernet: phy: Add KSZ8081 PHY Driver

Add Driver for KSZ8081 Ethernet PHY. The Generic MII Driver
is not sufficient to use for this PHY chip which has special
vendor implemented behaviors.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
dts: bindings: Add binding for KSZ8081 PHY

Add DT Binding for Microchip KSZ8081 PHY

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: ethernet: Add NXP ENET Driver

Add driver for NXP ENET which is a rework of the old
eth_mcux.c driver which had become unmaintainable due to
fundamental problems with the lack of PHY abstraction.

eth_mcux.c and the corresponding compatible nxp,kinetis-ethernet
will be deprecated and this new driver will be supported instead.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
soc: rt10xx: enable phy clock with new driver

when using either old or new driver for nxp enet,
enable the phy clock

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
dts: bindings: Add NXP ENET PTP binding

Add binding for NXP ENET PTP Clock device

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: ptp_clock: Add init priority Kconfig

Add ptp clock driver init priority kconfig

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: ptp_clock: Add NXP ENET PTP Clock Driver

Add Driver for NXP ENET PTP Clock device

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: clock_control_mcux_ccm: Add ENET PLL clk

Add subsys value for ENET PLL / ENET Ref clk to CCM Driver

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: eth_nxp_enet: Support PTP

Support PTP functionality in NXP ENET MAC driver

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
soc: nxp: rt10xx: Increase workqueue size for enet

Increase the size of the system workqueue stack if using
nxp ethernet driver

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
boards: mimxrt10xx_evk: Add new enet overlay

Add overlay to show how to use new experimental nxp enet driver

Also make any eth_mcux related kconfigs dependent on eth_mcux in the
boards' defconfigs.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

mmahadevan108 commented on dts/bindings/ethernet/phy_microchip_ksz8081.yaml:
```diff
@@ -0,0 +1,28 @@
+include: ethernet-phy.yaml
+
+properties:
+  mc,interface:
```
Should this property reside in `ethernet-phy.yaml`?

decsny replied:
yes, this phy can be configured for either mii or rmii interface

decsny replied:
I think I misread this comment, I think I know what you meant now, maybe this property can be generic and used by the enet driver since it has a phandle reference to the phy

mmahadevan108 commented on dts/bindings/ethernet/nxp,enet-mac.yaml:
```diff
@@ -0,0 +1,32 @@
+      - "mii"
+      - "rmii"
+    description: |
+      The interface that the PHY is connected to the ENET
```
Can we avoid this property and use what is provide in the phy node. Seems like the user will have to specify the same value in 2 places.

decsny replied:
no, it needs to be in both places, the phy and ethernet IP are independent devices

mmahadevan108 replied:
So they could be different?

decsny replied:
it wouldn't make sense for them to be different but both devices need to be configured for the correct interface. The fact that both the ethernet controller and the phy in this case have the functionality to configure the interface is just a coincidence. this ethernet controller could be used with a phy that is not configurable and needs to choose the right one, and the phy could be used with an ethernet controller that is not configurable and needs to be set up for the right interface. both of these bindings need this property. in the case of these EVKs, they are modular parts of the same system, but that is not the same as being the same device

decsny replied:
to align with linux it has to be a property of the ethernet controller (MAC), to align with what zephyr needs as far as initialization flow, it has to also be on this specific phy binding, because the phy cannot get information from the MAC node in this case without some complicated runtime callbacks, because there would be a circular reference in DT which is not allowed

mmahadevan108 commented on dts/bindings/ethernet/nxp,enet-mac.yaml:
```diff
@@ -0,0 +1,32 @@
+  nxp,fixed-link:
+    type: boolean
+    description: |
+      The ethernet instance operates with a fixed link phy.
```
Same comment as above.

decsny replied:
I don't know which comment you meant, but this property is not even used, I will just remove it, it's there by mistake I think, because there is now a fixed-link phy binding

mmahadevan108 commented:
Please update `west.yml` to pick the SHA

Commit:
dts: bindings: ethernet-controller: Add phy mode

Add a property to the ethernet controller binding
indicating what type of connection the MAC has with
the PHY device.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

decsny commented:
@mmahadevan108 update of your comments: https://github.com/zephyrproject-rtos/zephyr/compare/0a0c9fd0b82a9aaf0ea33f991f10a5bca8fdd0d4..d191b5d1b500ca71e39f68c84b42bdf58582b80e

manuargue commented on dts/bindings/ethernet/ethernet-phy.yaml:
```diff
@@ -25,3 +25,9 @@ properties:
       - "10BASE-T Full-Duplex"
       - "100BASE-T Half-Duplex"
       - "100BASE-T Full-Duplex"
+  interface:
```
it makes more sense to me to keep it in the eth driver because is the MAC which needs this information to most likely setup the interface to the PHY. And now that we are at it, we should probably make it part of `ethernet-controller` as `phy-connection-type`. See Linux binding as a reference: https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/net/ethernet-controller.yaml#L57
Most of the other eth bindings in Zephyr are already using `phy-connection-type`.

manuargue commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -0,0 +1,1045 @@
+	if (!state->is_up) {
+		net_eth_carrier_off(data->iface);
+		/* If link went down, keep trying to re-establish it */
+		while (phy_get_link_state(phy, &new_state) || !new_state.is_up) {
```
is this really necessary? `phy_mc_ksz8081_monitor_work_handler` is periodically checking for the link status. I'd suggest to instead bring the carrier down/up according to the link status passed to this cb.

decsny replied:
I am not sure what you mean with this, do you mean the phy driver should try to re-establish? or do you mean no reset is necessary and just wait more patiently?

manuargue replied:
If I understand correctly, the phy driver has delayable task monitoring the status of the link every `CONFIG_PHY_MONITOR_PERIOD` ms and on each link state change it will execute `nxp_enet_phy_cb` 
https://github.com/zephyrproject-rtos/zephyr/pull/62833/files#diff-8264e34cf3eb87668d72eefdb8154c2a3d452eea4e666168e8f43e01d9795306R374

When the link goes down, this callback will start resetting and re-configuring the phy and polling for the link status, I believe at the same time as the phy monitor task continue running. My suggestion is to let only the phy driver to monitor the link status at a fixed rate, and inform the enet driver when there's a link state change to do the needed re-configurations and notifications of the carrier status to the networking subsys. Then on this callback:
```c
if (state->is_up) {
    // perform any reconfiguration needed
    net_eth_carrier_on(data->iface);
} else {
    net_eth_carrier_off(data->iface);
}
```
For example, https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/ethernet/eth_sam_gmac.c#L1831-L1860
Perhaps I'm missing some piece to understand why this polling is needed.

decsny replied:
oh, I understand what you mean now, yes I agree, this was a mistake, since this callback function is already periodic it doesnt make sense to nest more periods, I will change it

manuargue commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -0,0 +1,1045 @@
+	const struct device *dev = eth_dev;
+	struct nxp_enet_mac_data *data = dev->data;
+
+	if (!state->is_up) {
```
if there's a change in speed/duplex mode, does the MAC needs to be reconfigured or it happens in-band? e.g. link was up at 100Mbps, goes down, goes up at 10Mbps

decsny replied:
I'm not 100% sure but I don't see a way from the RM to configure the MAC for a specific duplex or speed, but I have not been able to test this

decsny commented on drivers/mdio/mdio_nxp_enet.c:
```diff
@@ -0,0 +1,300 @@
+static void nxp_enet_mdio_bus_fn(const struct device *dev)
+{
+	/*
+	 * MDIO bus device is actually part of ethernet device, and
```
no, mdio must be initialized first, and the priorities are right

decsny commented on drivers/mdio/mdio_nxp_enet.c:
```diff
@@ -0,0 +1,300 @@
+#include <zephyr/sys_clock.h>
+#include <soc.h>
+
+/* Target MDC frequency 2.5 kHz */
```
looks like this is a typo, this is calculating MHz,

if interested here is a graph of the register value (5 bits 0-32) compared to the internal enet clock frequency used for mdc when using this calculation:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/110566814/d88738a1-787b-4b65-9038-717e99eb887a)

it always gives a value of at least 1 for the register

and corresponding graph of resulting mdc frequency:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/110566814/5e87cdac-e05e-406d-8b4e-1da29af143b5)

This equation comes from the SDK. I didn't try to come to a more optimal way of doing this, since the register value doesn't simply set a frequency but is relative to another clock, it is a bit complicated and I think the SDK probably spent a lot of time on this already. But it seems true that this will never result in the full 2.5 MHz for the MDC

manuargue replied:
thanks for the details. Yes I understand that not always will be able to use the maximum frequency, but the target frequency caught being 2.5 KHz caught my eyes. If it's a typo in your code comments can it be fixed then?

decsny replied:
yes I fixed it

decsny commented on drivers/mdio/mdio_nxp_enet.c:
```diff
@@ -0,0 +1,300 @@
+}
+
+/* MDIO Write API implementation */
+static int nxp_enet_mdio_write(const struct device *dev,
```
agree

jukkar reviewed (dismissed):
The networking code (seems to be mostly taken from the old driver) look reasonable so +1 for those changes.

pdgendt commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -0,0 +1,1046 @@
+		/* Free the TX buffer. */
+		k_sem_give(&data->tx_buf_sem);
+		break;
+	case kENET_ErrEvent:
+		/* Error event: BABR/BABT/EBERR/LC/RL/UN/PLR.  */
+		break;
```
Will `eth_callback` get called in this case when trying to send a frame? If that would be the case, the `tx_buf_sem` will never be released.
Or does the low level driver guarantee to call `eth_callback` with `kENET_TxEvent`?

decsny replied:
I will look into this further but in this case, this is not new code, I am only moving it to a different file. I think the answer depends on the hardware behavior which I cannot find documentation on. I am hesitant to just give this semaphore because this error interrupt is also used for many RX error situations, and I do not know if the MAC hardware will continue trying to send a TX frame if for example the FIFO is underrun which would cause this interrupt, so the answer of whether to give this semaphore here depends I think on hardware behavior which I am having trouble finding documentation for. If this is how it was in the old driver I think it is okay to leave it as is for now since this PR is not adding this as a new concept, but I will keep looking into this if I have time.

pdgendt commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -0,0 +1,1046 @@
+	/* Read network packet from upper layer into frame buffer */
+	if (net_pkt_read(pkt, data->tx_frame_buf, total_len)) {
+		k_mutex_unlock(&data->tx_frame_buf_mutex);
+		return -EIO;
```
The semaphore `&data->tx_buf_sem` has to be given back in this case.

pdgendt commented on dts/bindings/ethernet/microchip,ksz8081.yaml:
```diff
@@ -0,0 +1,27 @@
+include: ethernet-phy.yaml
+
+properties:
+  mc,reset-gpio:
```
Not sure about the `mc,` prefixes as these are common properties for all devices?

pdgendt commented on drivers/ethernet/phy/phy_microchip_ksz8081.c:
```diff
@@ -0,0 +1,460 @@
+	ret = phy_mc_ksz8081_read(dev, MII_BMSR, &bmsr);
+	if (ret) {
+		LOG_ERR("Error reading phy (%d) basic status register", config->addr);
+		return ret;
```
You need to unlock the mutex before returning.

AllemanMa commented on drivers/ethernet/phy/phy_microchip_ksz8081.c:
```diff
@@ -0,0 +1,460 @@
+
+	phy_mc_ksz8081_get_link(dev, &data->state);
+
+	data->cb(dev, &data->state, data->cb_data);
```
A bus fault occurs when `data->cb` is called here since `nxp_enet_phy_cb` calls `net_eth_carrier_on `which relies on `data->iface`. However iface seems to be not initialized yet. Default CONFIG_ETH_INIT_PRIORITY=80 and default CONFIG_NET_INIT_PRIO=90. Shouldn't we check if `data->iface` is initialized before calling `data->cb`?

decsny replied:
thanks, fixed

AllemanMa commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -0,0 +1,1046 @@
+				struct phy_link_state *state,
+				void *eth_dev)
+{
+	struct phy_link_state new_state;
```
Is there a reason why new_state remains unused?

decsny replied:
removed in latest push

pdgendt commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -0,0 +1,1046 @@
+	int ret = 0;
+
+	/* Reset the PHY */
+	ret = phy_write(phy, MII_BMCR, MII_BMCR_RESET);
```
As per the documentation of a the PHY LAN8720A:

> A Software reset is activated by setting the Soft Reset bit of the Basic Control Register to “1”. All registers bits, except those indicated as “NASR” in the register definitions, are cleared by a Software reset. The Soft Reset bit is self-clearing.
Per the IEEE 802.3u standard, clause 22 (22.2.4.1.1) the reset process will be completed within 0.5s from the setting
of this bit.

I don't think you can configure immediately after this call.
Perhaps we need to add a `reset` function to the PHY API?

AllemanMa commented on drivers/ethernet/phy/phy_microchip_ksz8081.c:
```diff
@@ -0,0 +1,459 @@
+	/* (re)do autonegotiation */
+	ret = phy_mc_ksz8081_autonegotiate(dev);
+	if (ret) {
+		LOG_ERR("Error in autonegotiation");
```
Unlock mutex before returning

AllemanMa commented on drivers/ethernet/phy/phy_microchip_ksz8081.c:
```diff
@@ -0,0 +1,459 @@
+	/* Software MDIO reset */
+	bmcr |= MII_BMCR_RESET;
+
+	ret = phy_mc_ksz8081_write(dev, MII_BMCR, bmcr);
```
Is it necessary to perform a software reset right after a hardware reset? Since `drivers/ethernet/eth_nxp_enet.c:571` performs a busy wait after a software reset, it should be done here as well if the software reset is still required ofcourse.

decsny replied:
I think this was left in by accident, I will remove it, it is not necessary, thanks

AllemanMa commented on drivers/mdio/mdio_nxp_enet.c:
```diff
@@ -0,0 +1,297 @@
+					(NSEC_PER_SEC / enet_module_clock_rate) - 1;
+	uint32_t mscr = ENET_MSCR_MII_SPEED(mii_speed) | ENET_MSCR_HOLDTIME(holdtime) |
+			(config->disable_preamble ? ENET_MSCR_DIS_PRE_MASK : 0);
+	config->base->MSCR = mscr;
```
Is there no need for lock and unlock the mutex while writing to the MSCR register?

decsny replied:
no, the mutex is for bus operation, which is initiated by functions part of the public api, so needs a mutex. this function just is called once for the module and is a private driver api, and even if multiple threads somehow call it (which should not happen) it doesn't matter because this is the only value ever written to this register.

AllemanMa commented on drivers/ptp_clock/ptp_clock_nxp_enet.c:
```diff
@@ -0,0 +1,268 @@
+	} else {
+		key = irq_lock();
+		if (config->base->ATPER != NSEC_PER_SEC) {
+			ret = -EBUSY;
```
Unlock irq before returning.

decsny commented:
twister failures addressed by https://github.com/zephyrproject-rtos/zephyr/pull/65323

pdgendt commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -0,0 +1,1053 @@
+#endif
+}
+
+#if defined(CONFIG_NET_IPV6)
+static void net_if_mcast_cb(struct net_if *iface,
+			    const struct net_addr *addr,
+			    bool is_joined)
+{
+	const struct device *dev = net_if_get_device(iface);
+	const struct nxp_enet_mac_config *config = dev->config;
+	struct net_eth_addr mac_addr;
+
+	if (addr->family != AF_INET6) {
+		return;
+	}
+
+	net_eth_ipv6_mcast_to_mac_addr(&addr->in6_addr, &mac_addr);
+
+	if (is_joined) {
+		ENET_AddMulticastGroup(config->base, mac_addr.addr);
+	} else {
+		ENET_LeaveMulticastGroup(config->base, mac_addr.addr);
+	}
+}
+#endif /* CONFIG_NET_IPV6 */
```
See #65232 and add this to the PR

IVandeVeire replied:
Please be aware of https://github.com/zephyrproject-rtos/zephyr/pull/65550

pdgendt commented on drivers/ethernet/Kconfig.nxp_enet:
```diff
@@ -0,0 +1,52 @@
+	  Place the hardware DMA buffers into DTCM for better
+	  networking performance.
+
+config ETH_NXP_ENET_HW_ACCELERATION
+	bool "Hardware acceleration for ethernet checksums"
```
Seems like this should be enabled by default?

decsny replied:
no, it doesn't work with ipv6

pdgendt commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -0,0 +1,1050 @@
+		  data->mac_addr,
+		  enet_module_clock_rate);
+
+#if defined(CONFIG_MDNS_RESPONDER) || defined(CONFIG_MDNS_RESOLVER)
+	ENET_AddMulticastGroup(config->base, mdns_multicast);
```
This should no longer be necessary to add here, with the changes in `net_if_mcast_cb`.

pdgendt commented on drivers/ethernet/phy/phy_microchip_ksz8081.c:
```diff
@@ -0,0 +1,444 @@
+
+	rc = phy_mc_ksz8081_get_link(dev, &state);
+
+	if (rc == 0 && memcmp(&state, &data->state, sizeof(struct phy_link_state))) {
```
Nit
```suggestion
	if (rc == 0 && memcmp(&state, &data->state, sizeof(struct phy_link_state)) != 0) {
```

pdgendt commented on drivers/ethernet/phy/phy_microchip_ksz8081.c:
```diff
@@ -0,0 +1,444 @@
+	}
+
+	/* TODO change this to GPIO interrupt driven */
+	k_work_reschedule(&data->phy_monitor_work,
+				K_MSEC(CONFIG_PHY_MONITOR_PERIOD));
```
Nit

```suggestion
	k_work_reschedule(&data->phy_monitor_work, K_MSEC(CONFIG_PHY_MONITOR_PERIOD));
```

pdgendt commented on drivers/mdio/mdio_nxp_enet.c:
```diff
@@ -0,0 +1,297 @@
+
+	/* This function will not make sense from IRQ context */
+	if (k_is_in_isr()) {
+		return -ECANCELED;
```
```suggestion
		return -EWOULDBLOCK;
```

pdgendt commented on drivers/ethernet/Kconfig.nxp_enet:
```diff
@@ -0,0 +1,52 @@
+
+if ETH_NXP_ENET
+
+config ETH_NXP_ENET_HW_ACCELERATION
+	bool "Hardware acceleration"
+	help
+	  Enable hardware acceleration for the following:
+	  - IPv4, UDP and TCP checksum (both Rx and Tx)
```
Add a dependency if it only works for IPv4.

```suggestion
config ETH_NXP_ENET_HW_ACCELERATION
	bool "Hardware acceleration"
	default y
	depends on !NET_IPV6
	help
	  Enable hardware acceleration for the following:
	  - IPv4, UDP and TCP checksum (both Rx and Tx)
```

jukkar replied:
But if you have both IPv4 and IPv6 enabled, then the HW acceleration would be disable for IPv4 too. Whatever check there is, it should be done at runtime.

pdgendt commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -0,0 +1,1047 @@
+	const struct nxp_enet_mac_config *config = dev->config;
+	enet_config_t enet_config;
+	uint32_t enet_module_clock_rate;
+#if defined(CONFIG_MDNS_RESPONDER) || defined(CONFIG_MDNS_RESOLVER)
+	/* standard multicast MAC address */
+	uint8_t mdns_multicast[6] = { 0x01, 0x00, 0x5E, 0x00, 0x00, 0xFB };
+#endif
```
Not used anymore

decsny commented:
@manuargue can you check if your block is addressed

manuargue reviewed (dismissed):
My comments addressed. Nice job!

danieldegrasse commented on include/zephyr/drivers/ethernet/eth_nxp_enet.h:
```diff
@@ -0,0 +1,40 @@
+ * Interrupt enable: The driver's relevant interrupt was enabled in NVIC
+ */
+enum nxp_enet_callback_reason {
+	nxp_enet_module_reset,
```
Nit- the convention is typically to use CAPS for a enum constant in Zephyr, right?

danieldegrasse reviewed (approved):
Couple of nits, but driver looks good to me. Verified it is working on an RT1060 EVK.

decsny commented on drivers/ethernet/phy/phy_microchip_ksz8081.c:
```diff
@@ -0,0 +1,443 @@
+ *
+ * Inspiration from phy_mii.c, which is:
+ * Copyright (c) 2021 IP-Logix Inc.
+ * Copyright 2022 NXP
```
this is intentional

danieldegrasse replied:
the two NXP copyright statements?

talih0 commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -0,0 +1,860 @@
+
+extern void nxp_enet_mdio_callback(const struct device *mdio_dev,
+				enum nxp_enet_callback_reason event);
+
```
I think this is already included in one of header you created.

talih0 commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -155,6 +166,86 @@ static void net_if_mcast_cb(struct net_if *iface,
+	struct net_pkt *pkt;
+
+	pkt = frameinfo->context;
+	if (pkt && atomic_get(&pkt->atomic_ref) > 0) {
```
Should the minimum ref count be actually 2?

If the reference count == 1, then the packet will be freed by net_pkt_unref() on line 235. That means whatever we push into the queue by net_if_add_tx_timestamp() is no longer valid.

talih0 commented on drivers/ptp_clock/Kconfig:
```diff
@@ -5,3 +5,13 @@ config PTP_CLOCK
+	int "Init priority"
+	default 75
+	help
+	  PTP Clock device driver initialization priority
```
We should update the other ethernet with ptp to use this.

decsny commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -0,0 +1,860 @@
+		return ret;
+	}
+
+	return ret;
```
my opinion on this is that it is easy to forget to add if another call is in the future added later in the code that affects ret, which can cause a bug that might not immediately appear, and the compiler will optimize this anyways, so I prefer this. I get this nit a lot

decsny commented on drivers/mdio/mdio_nxp_enet.c:
```diff
@@ -0,0 +1,297 @@
+		 * ethernet driver has not initiaized, just do a busy wait
+		 */
+		k_busy_wait(USEC_PER_MSEC * config->timeout);
+		if (base->EIR && ENET_EIR_MII_MASK == ENET_EIR_MII_MASK) {
```
this was fixed in #65999
